Architecture improvement

DmitryPogrebnoy opened the pull request:
Переименовал CoinClient в Coin, BlockchainListener в Explorer и наследуемые от них классы соответственно. Перетащил из arch ветки Coins класс. Немного изменил конфиги.

DmitryPogrebnoy committed:
Architecture improvement
Rename CoinClient -> Coin
Rename BlockchainListener -> Explorer
Separate Coins class
Improve configs
Code style

DmitryPogrebnoy committed:
Merge branch 'master' into BPS-20

# Conflicts:
#	src/main/kotlin/dsx/bps/config/currencies/BtcConfig.kt
#	src/main/kotlin/dsx/bps/config/currencies/TrxConfig.kt
#	src/main/kotlin/dsx/bps/config/currencies/XrpConfig.kt
#	src/main/kotlin/dsx/bps/core/BlockchainPaymentSystemManager.kt
#	src/main/kotlin/dsx/bps/crypto/btc/BtcCoin.kt
#	src/main/kotlin/dsx/bps/crypto/trx/TrxCoin.kt
#	src/main/kotlin/dsx/bps/crypto/xrp/XrpCoin.kt
#	src/test/kotlin/dsx/bps/core/BlockchainPaymentSystemManagerUnitTest.kt
#	src/test/kotlin/dsx/bps/crypto/btc/BtcClientUnitTest.kt
#	src/test/kotlin/dsx/bps/crypto/trx/TrxClientUnitTest.kt
#	src/test/kotlin/dsx/bps/crypto/xrp/XrpClientUnitTest.kt

DmitryPogrebnoy committed:
Separate connection configuration

skoret commented on src/main/kotlin/dsx/bps/config/currencies/BtcConfig.kt:
```diff
@@ -3,10 +3,16 @@ package dsx.bps.config.currencies
 import com.uchuhimo.konf.ConfigSpec
 
 object BtcConfig: ConfigSpec("btcConfig") {
```
коли стараемя сохранить простоту и последовательность в нейминге, то в "btcConfig" конфиг выглядит лишним

DmitryPogrebnoy replied:
"btcConfig" -> "BTC" и аналогичные так же

skoret commented on src/main/kotlin/dsx/bps/core/BlockchainPaymentSystemManager.kt:
```diff
@@ -22,34 +20,17 @@ class BlockchainPaymentSystemManager {
-        coins = mutableCoinsMap.toMap()
-
-        val emitters = coins.values.map { it.getTxEmitter() }
+        val emitters = coins.coins.values.map { it.getTxEmitter() }
```
мб добавить в Coins метод для получения эммитеров (всех или по конкретной валюте)?
коинс.коинс смущает немного

DmitryPogrebnoy replied:
Добавил в Coins метод получения всех эмиттеров. Получение эмиттера конкретной криптовалюты пока негде использовать, как понадобиться так и реализую.

skoret commented on src/main/kotlin/dsx/bps/crypto/btc/BtcCoin.kt:
```diff
@@ -53,34 +51,33 @@ class BtcClient: CoinClient {
             addSpec(BtcConfig)
```
initConfig используется здесь (и в остальных коинах) только чтобы попасть в with()
в конструкторе Coins похожая инициализация конфига выглядит лучше

- [x] поправить лишние переменные в коинах

DmitryPogrebnoy replied:
Поправил

skoret commented on src/main/kotlin/dsx/bps/crypto/btc/BtcCoin.kt:
```diff
@@ -53,34 +51,33 @@ class BtcClient: CoinClient {
-        val pass = config[BtcConfig.password]
-        val host = config[BtcConfig.host]
-        val port = config[BtcConfig.port]
+        val user = config[BtcConfig.Coin.user]
```
конструктор, принимающий путь к конфигу отличается от конструктора, принимающего объект конфига только тем, что подгружает файл в объект, а дальше логика одинаковая
причём все коин-клиенты грузят конфиг из файла одинаково

предлагаю в Coin добавить статик-метод, который по пути и спеке будет подгружать конфиг, валидировать его и возвращать
тогда конструктор, принимающий путь к конфигу, будет просто перенаправлять вызов в конструктор с подгруженным конфигом

- [x] переиспользовать конструктор с конфигом

DmitryPogrebnoy replied:
Выкинул конструктор конкретного Coin'a по пути до конфига т.к. он нигде не используется. Убрал проверку конфига в конструкторе конкретного Coin'a по конфигу. Перенес логику CoinFactory в CoinsManager, кажется это лучше отвечает принципу инкапсуляции.

skoret replied:
решение добавить в Currency метод по созданию соответствующего коин-клиента хорошо, равно как и фабрика, тем, что в коин-менеджере мы создаём любой клиент одной строчкой и добавление новых валют не заставит нас что-то менять в здесь
 
но сами currency ходят вообще по всей системе, в том числе появляются в invoice/paymentProcessor'ах, которые вроде как не должны знать про CoinClient'ов
а теперь в теории они могут создавать новые инстансы коин-клиентов, передав в currency спеку. то бишь модули начали пересекаться, инкапсуляция ломается

фабрика хоть и с одним статик-методом, но аналогично абстрактному Currency.createCoin обязывала разработчика добавить в неё пункт с созданием нового клиента, потому что там был switch (простите, when) по енуму
Currency больше про данные, чем про какую-то функциональность, — это идентификатор валюты

DmitryPogrebnoy replied:
То что currency ходят по всей системе настораживает, но с вариантом фабрики ничего не мешает "выстрелить в ногу" - вызвать фабричный метод, т.к. в currency уже есть необходимый ConfigSpec. Кроме того вызов фабричной функции происходит в одном месте и в обозримом будущем ничего не изменится.

Представим что добавляем новую валюту. Объясни, пожалуйста, разницу между тем что мы дописываем еще один вариант валюты в when в конструкторе CoinsManager с тем что мы дописываем еще один вариант валюты в when в фабричной функции.

Предлагаю альтернативный вариант(и кажется лучший) - можно перенести when в конструктор CoinsManager. Тем самым уберем синтетический фабричный класс и заберем у currency возможность создавать коины. А также по-прежненму необходимо будет добавить новую валюту в when в конструкторе CoinsManager иначе не скомпилится.

skoret replied:
это не должно настораживать, т.к. currency — идентификатор валюты. насторожит, если у нас для одной валюты будет несколько разных типов-идентификаторов
наличие в них спеков — вариант рабочий, но по названной тобою причине не особо идеологичный, да
с фабрикой выстрел в ногу хэндлится тем, что фабрика ограничена модулем crypto и до неё нельзя достать из core
размещение when в конструкторе аналогично методу фабрики
согласен, выглядит довольно подходяще

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/Coins.kt:
```diff
@@ -0,0 +1,84 @@
+import java.io.File
+import java.math.BigDecimal
+
+class Coins {
```
Это скорее CoinsRouter/Manager/etc

DmitryPogrebnoy replied:
Будет CoinsManager

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/Coins.kt:
```diff
@@ -0,0 +1,84 @@
+
+class Coins {
+
+    val coins: Map<Currency, Coin>
```
enabledCoins?

DmitryPogrebnoy replied:
Так и правда лучше)

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/Coins.kt:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    fun getTx(currency: Currency, txid: TxId): Tx {
+        val coin = getCoin(currency)
```
Нанопридирка, но кажется что вполне себе инлайн тут можно все дергать, а не разбивать на 2 строчки (это про все методы в этом классе)

DmitryPogrebnoy replied:
Действительно

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/Coins.kt:
```diff
@@ -0,0 +1,84 @@
+                try {
+                    coin.getTx(txid)
+                } catch (ex: Exception) {
+                    println(ex.message)
```
Тогда можно уж все вызовы обернуть в такой tryCatch

DmitryPogrebnoy replied:
Вот здесь я немного не понимаю.  @skoret что будет если запросить транзакцию по хешу которой нет в ноде? Код ответа будет 200 или другой?

skoret replied:
зависит от конкретной валюты
там даже не http-коды, а какой-нибудь json описывающий ошибку о ненайденной/несуществующей транзе
тут надо смотреть в CoinConnector.parseResponse
если на запрос вернулось не то, что мы ожидаем, сейчас бросается эксепшн

конкретно этот try-catch здесь, чтобы можно было лист хэшей замапить в лист вернувшихся транзакций, и не отвалиться на одной ненайденной

возможно, стоит пересмотреть решение выброса исключения в коннекторе на возвращение null

skoret replied:
кстати, здесь инлайн getCoin(currency) не подходит под общую схему
т.к. если мы во всех остальных методах отвалимся с эксепшеном из-за невключённой валюты, то тут вернём пустой список

DmitryPogrebnoy replied:
Действительно, ловится не нужная ошибка. Видимо стоит ловить только свои ошибки.
Почему нужно не отваливаться с ошибкой если хоть одна транзакция из списка не найдена?

skoret replied:
хотелось получить то, что получилось, а не отваливаться из-за одной пропавшей
возможно, обрабатывать здесь по-другому
но это уже на следующий раз

phil-dolgolev reviewed (commented):
С ренеймингами нужно быть аккуратнее

Если явно что-то плохо названо - то стоит менять, в противном случае обычно лучше забить, потому что сколько людей - столько мнений, и при постоянном ренейминге будет потом сложнее общаться на одном языке всем

DmitryPogrebnoy committed:
Minor improvement
Rename Coins -> CoinsManager
Rename Coins.coins -> CoinManager.enabledCoins
Delete unused constructor in each Coin class
Delete CoinFactory, move logic to CoinManager
Create function getAllEmitters in CoinsManager
Rename prefix in config file

DmitryPogrebnoy committed:
Move create coin logic to Currency

phil-dolgolev commented on src/main/kotlin/dsx/bps/core/BlockchainPaymentSystemManager.kt:
```diff
@@ -22,38 +20,18 @@ class BlockchainPaymentSystemManager {
-        emitter = Observable
-            .merge(emitters)
-            .observeOn(Schedulers.from(threadPool))
+        emitter = coinsManager.getAllEmitters().observeOn(Schedulers.from(threadPool))
```
Кстати, вопрос на подумать, а почему именно такое количество потоков выбирается ?)

DmitryPogrebnoy replied:
Берем все что доступно данному процессу)
Количество потоков возможно тоже стоит брать из конфига.

phil-dolgolev replied:
Это имеет смысл если на хостовой машине только твое приложение, а задачи CPU-heavy

IO-операции - вот совсем не CPU-heavy (а тут вроде много как раз IO) 

И кстати я понял, что не очень разобрался что здесь и на каком потоке обрабатывается, надо будет обсудить вживую наверное

phil-dolgolev commented on src/main/kotlin/dsx/bps/core/BlockchainPaymentSystemManager.kt:
```diff
@@ -71,40 +49,31 @@ class BlockchainPaymentSystemManager {
     }
 
-    constructor(coinClients: Map<Currency, CoinClient>, invoiceProcessor: InvoiceProcessor,
+    constructor(coinsMap: CoinsManager, invoiceProcessor: InvoiceProcessor,
```
наверное тут продолбался ренейминг

DmitryPogrebnoy replied:
Де, бывает

DmitryPogrebnoy committed:
Minor improvement
Add own exceptions
Add number of threads for Invoice observer to config

DmitryPogrebnoy commented:
Добавил свои базовые исключения. Добавил настройку количества потоков для invoice observer из конфига.
